> @asemjonovs - can you be a bit more specific about which toolchain(s) are giving you the error, or a log containing a compile error, or some reproducibility information?
> 
> Typically if a toolchain provides posix limits in the libc `limits.h` (and implements part of the POSIX API), it would not need to include Zephyr's `posix_limits.h`, which are fairly specific to Zephyr's implementation of the POSIX API. So if an external toolchain provides an implementation, it should use it's own `limits.h`.
> 
> Are you using a custom newlib by any chance?

We're using Coreboot SDK in our project.  
Here's the error that we are seeing
```
[redacted:rw] ccache /usr/bin/armv7m-cros-eabi-clang --target=armv7m-cros-eabi -DBORINGSSL_IMPLEMENTATION -DCHROMIUM_EC -DCONFIG_ZEPHYR -DKERNEL -DK_HEAP_MEM_POOL_SIZE=256 -DOPENSSL_NO_ASM -D_ANSI_SOURCE -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR__=1 -I/mnt/host/source/src/third_party/boringssl/include -I/mnt/host/source/src/platform/ec/build/zephyr/redacted/build-rw/zephyr/include/generated/zephyr -I/mnt/host/source/src/third_party/zephyr/main/include -I/mnt/host/source/src/platform/ec/build/zephyr/redacted/build-rw/zephyr/include/generated -I/mnt/host/source/src/third_party/zephyr/main/soc/nuvoton/npcx -I/mnt/host/source/src/third_party/zephyr/main/lib/libc/newlib/include -I/mnt/host/source/src/third_party/zephyr/main/lib/posix/options/getopt -I/mnt/host/source/src/third_party/zephyr/main/soc/nuvoton/npcx/common/. -I/mnt/host/source/src/third_party/zephyr/main/soc/nuvoton/npcx/common/npcxn/include -I/mnt/host/source/src/third_party/zephyr/main/soc/nuvoton/npcx/npcx9/. -I/mnt/host/source/src/third_party/zephyr/main/drivers -I/mnt/host/source/src/platform/ec/build/zephyr/redacted/include -I/mnt/host/source/src/platform/ec/build/zephyr/redacted/modules/ec/zephyr -I/mnt/host/source/src/platform/ec/build/zephyr/redacted/modules/ec/zephyr/drivers/usbc -I/mnt/host/source/src/third_party/zephyr/main/arch/arm/include -I/mnt/host/source/src/platform/ec/build/zephyr/redacted/build-rw/modules/ec/zephyr/include -I/mnt/host/source/src/third_party/zephyr/main/../../boringssl/include -I/mnt/host/source/src/platform/ec/third_party/boringssl/include -I/mnt/host/source/src/platform/ec/build/zephyr/redacted/modules/cmsis_6/CMSIS/Core/Include -I/mnt/host/source/src/third_party/zephyr/main/modules/cmsis_6/. -I/mnt/host/source/src/third_party/zephyr/main/include/zephyr/posix -isystem /mnt/host/source/src/third_party/zephyr/main/lib/libc/common/include -DSECTION_IS_RW -Werror -Wformat=2 -Wsign-compare -Wwrite-strings -Wvla -Wshadow -Wtype-limits -Wmissing-field-initializers -ggdb -Wall -fvisibility=hidden -fno-common -Wnewline-eof -fcolor-diagnostics -Wstring-concatenation -Wframe-larger-than=25344 -Wimplicit-fallthrough -Wmissing-prototypes -Wold-style-definition -Wstrict-prototypes -O3 -DNDEBUG -std=gnu11 -fno-strict-aliasing -Werror -Oz -flto=auto -imacros /mnt/host/source/src/platform/ec/build/zephyr/redacted/build-rw/zephyr/include/generated/zephyr/autoconf.h -fno-common -g -gdwarf-4 -fcolor-diagnostics -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfpu=fpv4-sp-d16 -mfloat-abi=hard -imacros /mnt/host/source/src/third_party/zephyr/main/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-unused-but-set-variable -Wno-typedef-redefinition -Wno-deprecated-non-prototype -Wno-pointer-sign -Wpointer-arith -Wno-main -Wno-unknown-warning-option -Wno-arm-interrupt-vfp-clobber -Wno-extra -Wno-self-assign -Wno-initializer-overrides -Wno-section -Wno-gnu -Wunused-variable -Werror=unused-variable -Werror=missing-braces -Werror=sometimes-uninitialized -Werror=unused-function -Werror=array-bounds -Werror=implicit-function-declaration -Werror=implicit-int -fno-PIC -fno-PIE -fno-asynchronous-unwind-tables -Wno-vla -fmacro-prefix-map=/mnt/host/source/src/platform/ec/zephyr/program/fpmcu=CMAKE_SOURCE_DIR -fmacro-prefix-map=/mnt/host/source/src/third_party/zephyr/main=ZEPHYR_BASE -ffunction-sections -fdata-sections -D_POSIX_THREADS -fmacro-prefix-map=/mnt/host/source/src/platform/ec/build/zephyr/redacted/modules/ec=EC_BASE -fmacro-prefix-map=/mnt/host/source/src/platform=EC_BASE -mno-unaligned-access -Wimplicit-fallthrough -Wint-conversion -std=c11 -ftrapv -D_POSIX_C_SOURCE=200809L -Wno-implicit-function-declaration -MD -MT modules/ec/third_party/boringssl/build/CMakeFiles/crypto.dir/crypto/asn1/a_bool.c.obj -MF modules/ec/third_party/boringssl/build/CMakeFiles/crypto.dir/crypto/asn1/a_bool.c.obj.d -o modules/ec/third_party/boringssl/build/CMakeFiles/crypto.dir/crypto/asn1/a_bool.c.obj -c /mnt/host/source/src/third_party/boringssl/crypto/asn1/a_bool.c
[redacted:rw] In file included from /mnt/host/source/src/third_party/boringssl/crypto/asn1/a_bool.c:57:
[redacted:rw] In file included from /mnt/host/source/src/third_party/boringssl/include/openssl/asn1.h:63:
[redacted:rw] In file included from /mnt/host/source/src/third_party/zephyr/main/include/zephyr/posix/time.h:59:
[redacted:rw] In file included from /mnt/host/source/src/third_party/zephyr/main/include/zephyr/kernel.h:17:
[redacted:rw] In file included from /mnt/host/source/src/third_party/zephyr/main/include/zephyr/kernel_includes.h:22:
[redacted:rw] In file included from /mnt/host/source/src/third_party/zephyr/main/lib/libc/newlib/include/limits.h:11:
[redacted:rw] /mnt/host/source/src/third_party/zephyr/main/include/zephyr/posix/posix_limits.h:83:9: error: 'ARG_MAX' macro redefined [-Werror,-Wmacro-redefined]
[redacted:rw]    83 | #define ARG_MAX                       _POSIX_ARG_MAX
[redacted:rw]       |         ^
[redacted:rw] /usr/armv7m-cros-eabi/include/sys/syslimits.h:36:9: note: previous definition is here
[redacted:rw]    36 | #define ARG_MAX                 65536   /* max bytes for an exec function */
[redacted:rw]       |         ^
[redacted:rw] In file included from /mnt/host/source/src/third_party/boringssl/crypto/asn1/a_bool.c:57:
[redacted:rw] In file included from /mnt/host/source/src/third_party/boringssl/include/openssl/asn1.h:63:
[redacted:rw] In file included from /mnt/host/source/src/third_party/zephyr/main/include/zephyr/posix/time.h:59:
[redacted:rw] In file included from /mnt/host/source/src/third_party/zephyr/main/include/zephyr/kernel.h:17:
[redacted:rw] In file included from /mnt/host/source/src/third_party/zephyr/main/include/zephyr/kernel_includes.h:22:
[redacted:rw] In file included from /mnt/host/source/src/third_party/zephyr/main/lib/libc/newlib/include/limits.h:11:
[redacted:rw] /mnt/host/source/src/third_party/zephyr/main/include/zephyr/posix/posix_limits.h:93:9: error: 'OPEN_MAX' macro redefined [-Werror,-Wmacro-redefined]
[redacted:rw]    93 | #define OPEN_MAX                      CONFIG_POSIX_OPEN_MAX
[redacted:rw]       |         ^
[redacted:rw] /usr/armv7m-cros-eabi/include/sys/syslimits.h:46:9: note: previous definition is here
[redacted:rw]    46 | #define OPEN_MAX                   64   /* max open files per process */
[redacted:rw]       |         ^
[redacted:rw] In file included from /mnt/host/source/src/third_party/boringssl/crypto/asn1/a_bool.c:57:
[redacted:rw] In file included from /mnt/host/source/src/third_party/boringssl/include/openssl/asn1.h:63:
[redacted:rw] In file included from /mnt/host/source/src/third_party/zephyr/main/include/zephyr/posix/time.h:59:
[redacted:rw] In file included from /mnt/host/source/src/third_party/zephyr/main/include/zephyr/kernel.h:17:
[redacted:rw] In file included from /mnt/host/source/src/third_party/zephyr/main/include/zephyr/kernel_includes.h:22:
[redacted:rw] In file included from /mnt/host/source/src/third_party/zephyr/main/lib/libc/newlib/include/limits.h:11:
[redacted:rw] /mnt/host/source/src/third_party/zephyr/main/include/zephyr/posix/posix_limits.h:96:9: error: 'PATH_MAX' macro redefined [-Werror,-Wmacro-redefined]
[redacted:rw]    96 | #define PATH_MAX                      _POSIX_PATH_MAX
[redacted:rw]       |         ^
[redacted:rw] /usr/armv7m-cros-eabi/include/sys/syslimits.h:48:9: note: previous definition is here
[redacted:rw]    48 | #define PATH_MAX                 1024   /* max bytes in pathname */
[redacted:rw]       |         ^
[redacted:rw] 3 errors generated.
```

--- a/doc/releases/release-notes-4.3.rst
+++ b/doc/releases/release-notes-4.3.rst
@@ -179,6 +179,10 @@ New APIs and options
     * :c:func:`zsock_listen` now implements the ``backlog`` parameter support. The TCP server
       socket will limit the number of pending incoming connections to that value.
 
+* Newlib
+
+  * :kconfig:option:`CONFIG_NEWLIB_LIBC_USE_POSIX_LIMITS_H`
+
 * Power management
 
    * :c:func:`pm_device_driver_deinit`

--- a/lib/libc/newlib/Kconfig
+++ b/lib/libc/newlib/Kconfig
@@ -70,4 +70,14 @@ config NEWLIB_LIBC_CUSTOM_SBRK
 	  Allow user to define custom version of the _sbrk function. Some application
 	  may need to use the remaining RAM for also other purposes than heap.
 
+config NEWLIB_LIBC_USE_POSIX_LIMITS_H
+	bool "Use Zephyr's posix_limits.h header"
+	default y
+	help
+	  The newlib provided by the Zephyr SDK does not include all mandatory POSIX limits.
+	  However, some external newlib-based toolchains do.
+
+	  Disable this option if your toolchain's newlib already includes all mandatory POSIX
+	  limits.
+
 endif # NEWLIB_LIBC

--- a/lib/libc/newlib/include/limits.h
+++ b/lib/libc/newlib/include/limits.h
@@ -8,6 +8,8 @@
 #define ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_LIMITS_H_
 
 #include_next <limits.h>
+#ifdef CONFIG_NEWLIB_LIBC_USE_POSIX_LIMITS_H
 #include <zephyr/posix/posix_limits.h>
+#endif
 
 #endif /* ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_LIMITS_H_ */

